identity_credential/src/revocation/status_list_2021/status_list.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2020-2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt::Display;
+
+use thiserror::Error;
+
+const DEFAULT_LIST_SIZE: usize = 16 * 1024;
+
+#[derive(Debug, Error, Eq, PartialEq)]
+/// [`std::error::Error`] type representing an invalidly encoded [`StatusList2021`]
+pub struct InvalidEncodedStatusList(String);
+
+impl Display for InvalidEncodedStatusList {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    write!(f, "{}", self.0)
+  }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+/// StatusList2021 data structure as described in [W3C's VC status list 2021](https://www.w3.org/TR/2023/WD-vc-status-list-20230427/)
+pub struct StatusList2021(Box<[u8]>);
+
+impl Default for StatusList2021 {
+  fn default() -> Self {
+    StatusList2021::new(DEFAULT_LIST_SIZE)
+  }
+}
+
+impl StatusList2021 {
+  /// Returns a new empty [`StatusList2021`] of `num_entries` entries
+  /// ## Notes:
+  /// - The actual length of the list might be slightly bigger
+  /// depending on the number of bytes the list will use.
+  /// - The specification suggests a list size of at least 16KBs
+  pub fn new(num_entries: usize) -> Self {
+    let size = num_entries / 8 + (num_entries % 8 != 0) as usize;
+    let store = vec![0; size];
+    StatusList2021(store.into_boxed_slice())
+  }
+  /// Returns the number of entries
+  #[allow(clippy::len_without_is_empty)]
+  pub const fn len(&self) -> usize {
+    self.0.len() * 8
+  }
+  /// Returns the status of the entry at `index` without bound checking
+  /// ## Panic:
+  /// * if `index` is greater than or equal to `self.len()`
+  pub const fn get_unchecked(&self, index: usize) -> bool {
+    let (i, offset) = Self::entry_index_to_store_index(index);
+    self.0[i] & (0b1000_0000 >> offset) != 0
+  }
+  /// Sets the status of the `index`-th entry to `value`
+  /// ## Panic:
+  /// * if `index` is greater than or equal to `self.len()`
+  pub fn set_unchecked(&mut self, index: usize, value: bool) {
+    let (i, offset) = Self::entry_index_to_store_index(index);
+    if value {
+      self.0[i] |= 0b1000_0000 >> offset
+    } else {
+      self.0[i] &= 0b0111_1111 >> offset
+    }
+  }
+  /// Returns the status of the `index`-th entry, if it exists
+  pub fn get(&self, index: usize) -> Option<bool> {
+    (index < self.len()).then_some(self.get_unchecked(index))
+  }
+  /// Sets the status fo the `index`-th entry to `value`
+  pub fn set(&mut self, index: usize, value: bool) {
```

Comment:
Better make this method fallible than silently not doing anything if index is out of bounds.